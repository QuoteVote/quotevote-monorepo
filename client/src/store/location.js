/**
 * Location Redux Store
 * Manages user's current location and permission state
 */

import { createSlice } from '@reduxjs/toolkit';
import { getCurrentPosition, checkPermission, GeoLocationError } from '../utils/geolocationService';

const initialState = {
  currentLocation: null, // { lat, lng }
  permissionState: 'prompt', // 'prompt' | 'granted' | 'denied' | 'unsupported'
  isLoading: false,
  error: null, // { type, message }
  lastUpdated: null,
};

const locationSlice = createSlice({
  name: 'location',
  initialState,
  reducers: {
    setLocation: (state, action) => {
      state.currentLocation = action.payload;
      state.lastUpdated = Date.now();
      state.error = null;
    },
    setPermissionState: (state, action) => {
      state.permissionState = action.payload;
    },
    setLoading: (state, action) => {
      state.isLoading = action.payload;
    },
    setError: (state, action) => {
      state.error = action.payload;
      state.isLoading = false;
    },
    clearError: (state) => {
      state.error = null;
    },
    clearLocation: (state) => {
      state.currentLocation = null;
      state.lastUpdated = null;
      state.error = null;
    },
    reset: () => initialState,
  },
});

export const {
  setLocation,
  setPermissionState,
  setLoading,
  setError,
  clearError,
  clearLocation,
  reset,
} = locationSlice.actions;

// Thunks

/**
 * Fetch user's current location
 * @param {object} options - Options passed to getCurrentPosition
 */
export const fetchLocation = (options = {}) => async (dispatch) => {
  dispatch(setLoading(true));
  dispatch(clearError());

  try {
    const location = await getCurrentPosition(options);
    dispatch(setLocation(location));
    dispatch(setPermissionState('granted'));
    return location;
  } catch (error) {
    dispatch(setError({
      type: error.type || GeoLocationError.UNKNOWN,
      message: error.message || 'Failed to get location',
    }));
    
    if (error.type === GeoLocationError.PERMISSION_DENIED) {
      dispatch(setPermissionState('denied'));
    }
    
    throw error;
  } finally {
    dispatch(setLoading(false));
  }
};

/**
 * Check and update permission state
 */
export const updatePermissionState = () => async (dispatch) => {
  try {
    const state = await checkPermission();
    dispatch(setPermissionState(state));
    return state;
  } catch (error) {
    console.error('Failed to check permission:', error);
    return 'prompt';
  }
};

// Selectors

export const selectLocation = (state) => state.location.currentLocation;
export const selectPermissionState = (state) => state.location.permissionState;
export const selectIsLoading = (state) => state.location.isLoading;
export const selectError = (state) => state.location.error;
export const selectLastUpdated = (state) => state.location.lastUpdated;

export const selectHasLocation = (state) => state.location.currentLocation !== null;
export const selectIsPermissionGranted = (state) => state.location.permissionState === 'granted';
export const selectIsPermissionDenied = (state) => state.location.permissionState === 'denied';

export default locationSlice.reducer;
