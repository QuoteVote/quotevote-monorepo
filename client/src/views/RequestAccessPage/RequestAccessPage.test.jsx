import React from 'react'
import { render } from '@testing-library/react'
import RequestAccessPage from './RequestAccessPage'

describe('RequestAccessPage', () => {
  it('renders correctly', () => {
    const { container } = render(<RequestAccessPage />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
/* eslint-disable no-undef */
/* eslint-disable react/jsx-no-undef */
import React from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import RequestAccessPage from './RequestAccessPage'
import store from '../../store/store'

const RequestAccessPageWrapper = () => (
  <ApolloProvider client={client}>
    <Provider store={store}>
      <RequestAccessPage />
    </Provider>
  </ApolloProvider>
)

describe('Request Access Page test -', () => {
  it('renders correctly', () => {
    const { container } = render(<RequestAccessPageWrapper />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
