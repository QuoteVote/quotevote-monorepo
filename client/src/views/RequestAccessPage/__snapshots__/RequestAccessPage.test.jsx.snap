// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Request Access Page test - > renders correctly 1`] = `
<div
  class="makeStyles-container-2"
>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column MuiGrid-spacing-xs-2 css-h6dotr-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
      style="justify-content: center; margin-bottom: 2rem;"
    >
      <img
        alt="logo"
        src="/assets/quote-vote-white.svg"
        style="width: 100%; height: auto;"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInputBase-colorPrimary makeStyles-input-41 css-mcfk8m-MuiInputBase-root-MuiInput-root"
      >
        <input
          class="MuiInputBase-input MuiInput-input css-1smhhcz-MuiInputBase-input-MuiInput-input"
          placeholder="Enter Email"
          type="text"
          value=""
        />
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary makeStyles-button-50 makeStyles-requestAccessBtn-42 css-1xjyx2h-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Request Invite
      </button>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
    >
      <div
        class="makeStyles-overlayContainer-43"
        style="max-width: 100%; margin: 3rem auto 0px auto;"
      >
        <div
          class="makeStyles-overlay-44"
        />
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-6 makeStyles-overlayContent-45 css-65mziv-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-0"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-13y79c5-MuiTypography-root"
                style="text-align: center;"
              >
                No Ads, No Algorithms
              </h5>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
                style="margin-bottom: 8px; text-align: left;"
              >
                There is no ranking, boosting, or personalization engine. You can't pay to be seen. Users seek to find quotes.
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
                style="text-align: left;"
              >
                Discovery is deliberate. Feeds are chronological. An experience of hunting and discovery, not passive scrolling.
              </p>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
          >
            <div
              class="MuiBox-root css-0"
            >
              <h5
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-13y79c5-MuiTypography-root"
                style="text-align: center;"
              >
                Open Source, Non Profit
              </h5>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
                style="margin-bottom: 8px; text-align: left;"
              >
                The platform is non-profit, open source, and donation-supported. You can't pay to be seen.
              </p>
              <p
                class="MuiTypography-root MuiTypography-body1 css-1ip8nm7-MuiTypography-root"
              >
                The only economic model is when users like what they experience, and give money via donations.
              </p>
            </div>
          </div>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary makeStyles-button-50 makeStyles-requestAccessBtn-42 css-1xjyx2h-MuiButtonBase-root-MuiButton-root"
        style="display: block; margin: 2rem auto 0px auto;"
        tabindex="0"
        type="button"
      >
        Read Our Mission
      </button>
    </div>
  </div>
</div>
`;
